arduino/seed_aura/seed_aura.ino

```diff
@@ -0,0 +1,186 @@
+  uint32_t pot_3 = readPot(PIN_POT_3);
+
+  // analog write to led 1
+  writeLed(PIN_LED_1, pot_1 * 0.66f + pot_2 * 0.33f);
+
+  // map pot_1 to presence
+  float presence = mapf(pot_1, 0.0f, 1023.0f, 0.0f, 1.0f);
+  // map pot_2 to gain
+  float gain = mapf(pot_2, 0.0f, 1023.0f, 0.0f, 1.8f);
+
+  if (sw_1 == LOW) {
+
+    // map pot_3 to freq
+    float freq = mapf(pot_3, 0.0f, 1023.0f, 30.0f, 3000.0f);
+
+    // analog write to led 2
+    writeLed(PIN_LED_2, pot_3 * 0.66f + pot_2 * 0.33f);
+
+    // set filter frequency
+    lpf_l.SetFreq(freq);
+    lpf_r.SetFreq(freq);
+  }
+  else if (sw_1 == HIGH) {
+
+    // map pot_3 to lfo rate
+    float lfoRate = mapf(pot_3, 0.0f, 1023.0f, MIN_LFO_RATE_HZ, MAX_LFO_RATE_HZ);
+
+    // update lfo rate
+    lfo_l.SetFreq(lfoRate);
+    lfo_r.SetFreq(lfoRate);
+    float v_l = lfo_l.Process();
+    float v_r = lfo_r.Process();
+
+    // analog write to led 2
+    writeLed(PIN_LED_2, mapf(v_l, -1.0f * DEPTH_HZ, DEPTH_HZ, 0.0f, 1.0f) * 1023.0f);
+
+    // set filter frequency
+    lpf_l.SetFreq(MIN_FREQ_HZ + (DEPTH_HZ / 2.0f) + v_l);
+    lpf_r.SetFreq(MIN_FREQ_HZ + (DEPTH_HZ / 2.0f) + v_r);
+  }
+
+  // apply gain, presence, soft clip
+  for (size_t i = 0; i < size; i++) {
+
+    float pre_l = in[0][i];
+    float post_l = pre_l * gain;
+    float inverted_l = -1.0f * post_l;
+    float filtered_l = lpf_l.Process(inverted_l);
+    out[0][i] = softClip(pre_l + filtered_l * (1.0f - presence));
+
+    float pre_r = in[1][i];
+    float post_r = pre_r * gain;
+    float inverted_r = -1.0f * post_r;
+    float filtered_r = lpf_r.Process(inverted_r);
+    out[1][i] = softClip(pre_r + filtered_r * (1.0f - presence));
+  }
+}
+
+void setup() {
+
+  // pull up switch digital pins
+  pinMode(PIN_SW_1, INPUT_PULLUP);
+  pinMode(PIN_FS_1, INPUT_PULLUP);
+
+  // initialize daisy hardware
+  seed = DAISY.init(DAISY_SEED, AUDIO_SR_96K);
+
+  float sampleRate = DAISY.get_samplerate();
+
+  // low pass filters
+  lpf_l.Init(sampleRate);
+  lpf_r.Init(sampleRate);
+
+  // triangle lfo
+  lfo_l.Init(sampleRate);
+  lfo_l.SetWaveform(Oscillator::WAVE_TRI);
+  lfo_l.SetAmp(DEPTH_HZ);
+  lfo_l.SetFreq(MIN_LFO_RATE_HZ);
+
+  // triangle lfo, out of phase
+  lfo_r.Init(sampleRate);
+  lfo_r.SetWaveform(Oscillator::WAVE_TRI);
+  lfo_r.SetAmp(DEPTH_HZ);
+  lfo_r.SetFreq(MIN_LFO_RATE_HZ);
+  lfo_r.PhaseAdd(0.5f);
+
+  // register audio callback
+  DAISY.begin(audio);
+}
+
+void loop() {}
+
+int readSwitch(int32_t pin) {
+  return digitalRead(pin);
+}
+
+uint32_t readPot(int32_t pin) {
+  return constrain(1023 - analogRead(pin), 0, 1023);
+}
+
+void writeLed(int32_t pin, int32_t value) {
+  analogWrite(pin, constrain(value / 4, 0, 255));
+}
+
+float softClip(float x) {
+  if (x < -3.0f) {
+    return -1.0f;
+  }
+  else if (x > 3.0f) {
+    return 1.0f;
+  }
+  return x * (27.0f + x * x) / (27.0f + 9.0f * x * x);
+}
+
+float mapf(float x, float in_min, float in_max, float out_min, float out_max) {
```

Comment:
There are fast versions of similar functions in
https://github.com/electro-smith/DaisySP/blob/master/Source/Utility/dsp.h#L40

Perhaps this could be improved?